### PR TITLE
feat!: correctly creating the data commitments when the window is changed

### PR DESCRIPTION
**Pull request by SweeXordious**
## Overview

This PR allows creating the data commitments correctly when the data commitments window changes. This change can be via an upgrade or a gov proposal. 

Closes https://github.com/celestiaorg/celestia-app/issues/1685

### CODE REVIEW

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:263]: This one is only used for testing. To simulate what happens in end blocker. Should it also be tested?

**Review by cmwaters — COMMENTED** [review]
I think creating a data commitment should not be predicated on:

```go
if ctx.BlockHeight() != 0 && ctx.BlockHeight()%int64(k.GetDataCommitmentWindowParam(ctx)) == 0 {
```

but

```go
if ctx.BlockHeight() != 0 && ctx.BlockHeight()-k.GetLastDataCommitment(ctx).EndBlock >= int64(k.GetDataCommitmentWindowParam(ctx)) {
```

**Comment by SweeXordious** [comment]
In case of a window change, for example from 400 to 199 at block 900,

```
if ctx.BlockHeight() != 0 && ctx.BlockHeight()%int64(k.GetDataCommitmentWindowParam(ctx)) == 0 {
```

Would give us ranges like: `[1, 400], [401, 800], [801, 995], [996, 1194]`. Which gives us an intermediate data commitment that has a custom range.

```
if ctx.BlockHeight() != 0 && ctx.BlockHeight()-k.GetLastDataCommitment(ctx).EndBlock >= int64(k.GetDataCommitmentWindowParam(ctx)) {
```
Would give us: `[1, 400], [401, 800], [801, 999], [1000, 1199]`, which wouldn't have the intermediate data commitment that has a custom range.

So that depends on whether we want to have an intermediate data commitment range or not

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:133]: Does this work? [90593fd](https://github.com/celestiaorg/celestia-app/pull/1695/commits/90593fd72a94445371786f6e49970704996e4d7c)

File: x/qgb/keeper/keeper_data_commitment.go
Comment (evan-forbes) [inline x/qgb/keeper/keeper_data_commitment.go:35]: will defer to author is they think this is simpler
```suggestion
		beginBlock = lastDCC.EndBlock + 1
		endBlock = lastDCC.EndBlock + dcWindow
```

File: x/qgb/keeper/keeper_data_commitment.go
Comment (evan-forbes) [inline x/qgb/keeper/keeper_data_commitment.go:28]: maybe add a doc just to clarify that this is only for the first data commitment

File: x/qgb/abci_test.go
Comment (evan-forbes) [inline x/qgb/abci_test.go:178]: as mentioned in a different reviewer's comment, parsing this is difficult because its difficult to understand what criteria is being checked. so we could perhaps add a few comments to elaborate on what we are looking for and why.

**Review by evan-forbes — COMMENTED** [review]
question: what happens if the data commitment gets changed to something so small that multiple data commitments are needed? for example, what if we go from 400 -> 5? 400-> 1?

mostly just doc stuff otherwise LGTM

**Comment by SweeXordious** [comment]
> what happens if the data commitment gets changed to something so small that multiple data commitments are needed? for example, what if we go from 400 -> 5? 400-> 1?

We will have an intermediary data commitment that will contain all those data commitments, then continue with the provided range. This comment has some details about it: https://github.com/celestiaorg/celestia-app/pull/1695#issuecomment-1534930915

So, for example if we go from 400 to 2 (1 shouldn't be possible, Also, we're setting the lower bound of windows to 101, so we shouldn't get into that case), and we activate the change at height 1000, we will have the following ranges: `[1, 400], [401, 800], [801, 1000], [1001, 1002], [1003, 1004]...`

**Comment by SweeXordious** [comment]
I guess I will document this in https://github.com/celestiaorg/celestia-app/tree/main/x/qgb#new-data-commitment-creation

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:178]: What do you think about this explanation? [90593fd](https://github.com/celestiaorg/celestia-app/pull/1695/commits/90593fd72a94445371786f6e49970704996e4d7c).
If not clear enough, I will inline comments with every line of the test. 
Let me know what you think is better.

File: x/qgb/keeper/keeper_data_commitment.go
Comment (SweeXordious) [reply x/qgb/keeper/keeper_data_commitment.go]: [36c4523](https://github.com/celestiaorg/celestia-app/pull/1695/commits/36c45239776af731f97e22bf0b35ec83a1ea2c98)

File: x/qgb/abci_test.go
Comment (rootulp) [reply x/qgb/abci_test.go:133]: This definitely helps. I wonder if it's also possible to modify the ranges to make them easier for humans to read. What if the start of the range was inclusive and the end of the range was exclusive. Then for a data commitment window of 400 that starts at `0`, we'd have:

0 - 400
400 - 800

and if we had to start at `1` because weirdness around the genesis block:

1 - 401
401 - 801

It's a bit easier to tell that these ranges are the correct size because end - start = data commitment window

Comment (rootulp) [reply x/qgb/abci_test.go:133]: A few ideas to make this test easier to read:

1. Split the long test into multiple test cases, potentially one per range
2. Extract a range struct that specifies a begin and end block for a data commitment window. Then assert that the expected range matches the result's data commitment `BeginBlock` / `EndBlock`

File: x/qgb/abci_test.go
Comment (rootulp) [inline x/qgb/abci_test.go:163]: [question] is there a rationale to use `101` instead of `100`? I would expect `100` to produce friendlier (to the human eye) ranges to verify

```suggestion
	newShrinkedDCWindow := 101
```

File: x/qgb/abci_test.go
Comment (rootulp) [inline x/qgb/abci_test.go:182]:
```suggestion
	newExpandedDCWindow := 500
```

File: x/qgb/keeper/keeper_data_commitment.go
Comment (rootulp) [reply x/qgb/keeper/keeper_data_commitment.go:14]: IMO describing that this function depends on the data commitment window is an implementation detail that doesn't need to be documented. Interested readers can read the function body. If we avoid documenting it then we don't need to update the godoc when implementation details change. So my proposal would be:

```
// GetCurrentDataCommitment creates the latest data commitment at the current height.
```

Although that makes me think that more than one data commitment can exist for a given height. Why is "latest" necessary?

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:163]: 100 can give ranges that are similar to the ones for 400, if the upgrade height is 1200 for example. That's why I used 101 so that we can see the difference.
Also, 101 is the lower bound for data commitment windows, but it's unrelated here.

File: x/qgb/keeper/keeper_data_commitment.go
Comment (SweeXordious) [reply x/qgb/keeper/keeper_data_commitment.go:14]: cool, will remove the comment.

> Although that makes me think that more than one data commitment can exist for a given height. Why is "latest" necessary?

Only one data commitment can exist for height. Also, `latest` means in here the last one that will be added

File: x/qgb/abci_test.go
Comment (rootulp) [reply x/qgb/abci_test.go:163]: > That's why I used 101

Makes sense, thanks!

> Also, 101 is the lower bound for data commitment windows, but it's unrelated here.

Where is this lower bound specified? Why is `101` the lower bound?

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:133]: > This definitely helps. I wonder if it's also possible to modify the ranges to make them easier for humans to read. What if the start of the range was inclusive and the end of the range was exclusive. Then for a data commitment window of 400 that starts at 0, we'd have

We definitely could. I am using the inclusive ranges because when we say: `beginBlock: 1, endBlock: 100`, we automatically assume that it's inclusive. And if it's not, we would need to add a comment for it everywhere we use it. Especially if we switch the using `firstBlock`, and `lastBlock` words. Thus, I kept using the inclusive ranges.


> A few ideas to make this test easier to read:
> Split the long test into multiple test cases, potentially one per range
> Extract a range struct that specifies a begin and end block for a data commitment window. Then assert that the expected range matches the result's data commitment BeginBlock / EndBlock

We could do that but I wanted to have the sequential execution of the test and asserting that the ranges are continuous (if it's the right word to use here). That's why I opted for a sequential list of actions.

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:163]: The lower bound is specified in here:
https://github.com/celestiaorg/celestia-app/blob/b2e8a72c76d40003acb93b988debe92cd110ae99/x/qgb/types/genesis.go#L54-L55

Now that I am checking, it's 100 :D 

The reason for that is if we start sending attestations for lower ranges, we will need to relay more transactions, which means spending more money. Also, since we're able to batch lots of blocks in the same data commitment, the more we batch, the easier it is to find the commitments and prove them.

File: x/qgb/abci_test.go
Comment (evan-forbes) [reply x/qgb/abci_test.go:133]: not blocking on it, but fwiw I'd +1 any efforts to make the test more readable. Its testing complex cases, which I know isn't easy to abstract or make readable, but it is currently very beefy

perhaps we could break into test cases? it looks like we're comparing begin, endblock, and nonce 4 separate times, could we keep those heights in a test case as expected values? then we're also setting the attestation request between each iteration, could that be part of the test input?

**Review by evan-forbes — DISMISSED** [review]
not blocking, but left some comments :slightly_smiling_face:

**Comment by SweeXordious** [comment]
Thanks :+1: I will try to refactor that test tomorrow and see.

Also, your opinion on this is valuable: https://github.com/celestiaorg/celestia-app/pull/1695#issuecomment-1534930915

Which do you think is better?

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go:133]: Is this better [eaf055c](https://github.com/celestiaorg/celestia-app/pull/1695/commits/eaf055c9b701f668ae6d66181cbbdd46326e6110) ?

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go]: Deleted here [8e57c11](https://github.com/celestiaorg/celestia-app/pull/1695/commits/8e57c11b46c9e432186a7681646e064db545c5bb)

File: x/qgb/abci_test.go
Comment (evan-forbes) [reply x/qgb/abci_test.go:133]: yes! thanks!

**Review by evan-forbes — APPROVED** [review]

File: x/qgb/abci_test.go
Comment (rootulp) [reply x/qgb/abci_test.go]: Thanks for sharing!

1. should the `100` be extracted to a constant?
2. I don't see why specifying a data commitment window of `99` would result in the error: `"invalid average EVM block time, too short for latency limitations"` because the data commitment window is not related to the average EVM block time. I would expect an error message like

```go
const MinimumDataCommitmentWindow = 100

func validateDataCommitmentWindow(i interface{}) error {
	val, ok := i.(uint64)
	if !ok {
		return fmt.Errorf("invalid parameter type: %T", i)
	} else if val < uint64(MinimumDataCommitmentWindow) {
		return fmt.Errorf("data commitment window %v must be >= minimum data commitment window %v", val, MinimumDataCommitmentWindow)
	}
	return nil
}
```

File: x/qgb/keeper/keeper_data_commitment.go
Comment (SweeXordious) [reply x/qgb/keeper/keeper_data_commitment.go:14]: to the state, this method will create a new data commitment and add it to the state, making it the latest data commitment

File: x/qgb/abci_test.go
Comment (SweeXordious) [reply x/qgb/abci_test.go]: > should the 100 be extracted to a constant?

Could be. But I don't see a reason for now why

> I don't see why specifying a data commitment window of 99 would result in the error: "invalid average EVM block time, too short for latency limitations" because the data commitment window is not related to the average EVM block time. I would expect an error message like

I'll get to it  and update the message https://github.com/celestiaorg/celestia-app/issues/1706, thanks

Comment (rootulp) [reply x/qgb/abci_test.go]: Yea sorry this comment is not blocking this PR. Good idea to create a separate issue. Resolving!

File: x/qgb/keeper/keeper_data_commitment.go
Comment (rootulp) [reply x/qgb/keeper/keeper_data_commitment.go:14]: [non-blocking] here's what I would expect the godoc to look like strictly based on the function name:

```go
// GetCurrentDataCommitment returns a data commitment for the current height.
func (k Keeper) GetCurrentDataCommitment(ctx sdk.Context) (types.DataCommitment, error) {
```

In the function body, I don't see where it adds a data commitment to the state. If it does indeed update state, would a more descriptive function be:

```go
// UpdateLatestDataCommitment creates a new data commitment for the current
// height, updates it in state, and returns it.
func (k Keeper) UpdateLatestDataCommitment(ctx sdk.Context) (types.DataCommitment, error) {
```

**Review by rootulp — APPROVED** [review]
Approving to unblock but TBH I still don't understand the ranges or how this fixes

**Comment by SweeXordious** [comment]
@rootulp The data commitment is set here:

https://github.com/celestiaorg/celestia-app/blob/main/x/qgb/abci.go#L28

Also, this shows I will need to write more documentation around that. Thanks for letting me know.

**Comment by rootulp** [comment]
Based on https://github.com/celestiaorg/celestia-app/blob/main/x/qgb/abci.go#L28 the function `GetCurrentDataCommitment` doesn't update the data commitment in state so we can use the more accurate:

```go
// GetCurrentDataCommitment returns a data commitment for the current height.
func (k Keeper) GetCurrentDataCommitment(ctx sdk.Context) (types.DataCommitment, error) {
```

**Comment by SweeXordious** [comment]
> Based on https://github.com/celestiaorg/celestia-app/blob/main/x/qgb/abci.go#L28 the function `GetCurrentDataCommitment` doesn't update the data commitment in state so we can use this option instead of the other option:
> 
> ```go
> // GetCurrentDataCommitment returns a data commitment for the current height.
> func (k Keeper) GetCurrentDataCommitment(ctx sdk.Context) (types.DataCommitment, error) {
> ```

Cool, opened issue and will update in a subsequent PR https://github.com/celestiaorg/celestia-app/issues/1710, thanks

**Comment by SweeXordious** [comment]
Merging, will address the documentation changes in subsequent PRs. Thanks a lot everyone for the thorough reviews :+1: :+1:

**Comment by cmwaters** [comment]
> So that depends on whether we want to have an intermediate data commitment range or not

Yeah the reasoning behind why I mentioned that was to specifically remove this intermediate commitment range. There should just be the old one and the new one. 

> Edit: The second way will not give us an intermediate data commitment range if the upgrade/gov proposal was done in the range [lastDataCommitment.EndBlock, lastDataCommitment.EndBlock + newWindow]. If not, it will behave similar to the first way and have an intermediate data commitment that has a custom range

I'm a bit confused by this, I thought this would remove the possibility of an intermediate data commitment range in all cases

**Comment by SweeXordious** [comment]
> I'm a bit confused by this, I thought this would remove the possibility of an intermediate data commitment range in all cases

Let's take for example an initial data commitment window of 500.
Then, the first ranges, up to height 1000, will be:
- `[1, 500]`
- `[501, 1000]`

Then, at block 1300, we make a gov proposal to change the window down to 100. Then, the following condition:
```
if ctx.BlockHeight() != 0 && ctx.BlockHeight()-k.GetLastDataCommitment(ctx).EndBlock >= int64(k.GetDataCommitmentWindowParam(ctx)) {
```
will be true at height 1300, and we will create a new data commitment as follows:

https://github.com/celestiaorg/celestia-app/blob/c08d1c690945b27c8341d91708419c872f3b69cd/x/qgb/keeper/keeper_data_commitment.go#L31-L36

Which will be: `[1001, 1300]`.

After that, we will have the following ranges: `[1301, 1400], [1401, 1500]...`

Thus, we will still have the intermediate range of `[1001, 1300]`. Unless I am missing your point.

If we don't want to have intermediate ranges, one way is to create multiple attestations at the same height. In the above case, at height 1300, we create the following attestations: `[1001, 1100], [1101, 1200], [1201, 1300]`.

**Comment by cmwaters** [comment]
> Which will be: [1001, 1300].

Won't it be [1001, 1100] if the `dcWindow` is 100?

**Comment by SweeXordious** [comment]
yes, that if the window change happens in height in the following `[1001, 1100]`. However, if it happens later, it will be the above

**Comment by SweeXordious** [comment]
This assumes that end blocker will run multiple times until we have all the attestations created. However, it's only ran once. 

This is what I meant by adding the possibility to creating multiple attestations per height if needed and we won't have the intermediate range. But that would require changing the implementation of end blocker.

**Comment by SweeXordious** [comment]
I will create an issue to investigate having this. However, I still don't why having an intermediate range is bad. Please let me know if there is something I'm missing

**Comment by evan-forbes** [comment]
> yes, that if the window change happens in height in the following [1001, 1100]. However, if it happens later, it will be the above

my understanding of the code was what @cmwaters said, and that's what appears to happen in the [test](https://github.com/celestiaorg/lazyledger-app/blob/e223493e3bb53a969b18297150a6de20202feaec/x/qgb/abci_test.go#L165-L173) if we modify it to the above scenario.

this is mainly why I asked [this question](https://github.com/celestiaorg/celestia-app/pull/1695#pullrequestreview-1413756839), because it looked like we could only create a single commitment per block, and we would only set a data commitment if the block height is a multiple of the window. Meaning that we might be in a scenario where it takes a long time for the data commitment to catch up to the block height.

https://github.com/celestiaorg/celestia-app/blob/e223493e3bb53a969b18297150a6de20202feaec/x/qgb/abci.go#L23

This was also why I thought https://github.com/celestiaorg/celestia-app/pull/1713 solved, was by setting to the min to 100, its reasonable that we would always catch up.

so... imo we should keep the code as is here, but then add the ability to create multiple data commitments to cover the edge case. 

perhaps we're missing something, and if so, @SweeXordious do you mind adding a test that demonstrates the intermediate range?

**Comment by SweeXordious** [comment]
> but then add the ability to create multiple data commitments to cover the edge case.

This is so we don't have the intermediate range, right?

Why is it not good to have an intermediate range that's used to catchup? 

Sure, will create a test and let you know

**Comment by cmwaters** [comment]
> However, I still don't why having an intermediate range is bad

I don't think it's that bad (I don't have a full view on things), it's just not correct. You've gone from one range to another so those would be the two ranges we expect posted but now there's a third range to "bridge the gap".

**Comment by SweeXordious** [comment]
Per sync, will create an issue to catchup on the data commitments. Good catch guys :+1: :+1:

**Pull request merged**
